### PR TITLE
[CBC Gem] Support 1080p shows

### PR DESCRIPTION
**Pull request by DavidSkrundz**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This adds a new check to pull out "secret" video formats such at 6Mb 1080p. The original video formats begin with `hls-` while the new ones begin with `sec-`. Here is sample output with the changes (requires a free CBC account and geolocation within Canada):

```
~ % yt-dlp "https://gem.cbc.ca/media/murdoch-mysteries/s01e01" --username '***' --password '***' -F
[gem.cbc.ca] murdoch-mysteries/s01e01: Downloading JSON metadata
[gem.cbc.ca] murdoch-mysteries/s01e01: Downloading JSON metadata
[gem.cbc.ca] murdoch-mysteries/s01e01: Downloading m3u8 information
[gem.cbc.ca] murdoch-mysteries/s01e01: Downloading secret XML
[info] Available formats for murdoch-mysteries/s01e01:
ID                              EXT RESOLUTION │   FILESIZE   TBR PROTO  │ VCODEC        VBR ACODEC     MORE INFO
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
hls-audio-English__Descriptive_ m4a            │                  m3u8_n │ audio only        mp4a.40.2  [eng] English (Descriptive)
hls-audio-English               m4a            │                  m3u8_n │ audio only        mp4a.40.2  [eng] English
sec-234-0                       mp4 416x234    │ ~ 32.34MiB   99k m3u8_n │ avc1.4d401f   99k video only
sec-234-1                       mp4 416x234    │ ~ 64.70MiB  199k m3u8_n │ avc1.4d401f  199k video only
sec-234-2                       mp4 416x234    │ ~129.39MiB  399k m3u8_n │ avc1.4d401f  399k video only
hls-621                         mp4 416x234    │ ~200.99MiB  621k m3u8_n │ avc1.42c00d  621k video only
sec-360-0                       mp4 640x360    │ ~194.09MiB  599k m3u8_n │ avc1.4d401f  599k video only
hls-825                         mp4 640x360    │ ~267.11MiB  825k m3u8_n │ avc1.42c01e  825k video only
sec-360-1                       mp4 640x360    │ ~320.33MiB  990k m3u8_n │ avc1.4d401f  990k video only
hls-1224                        mp4 640x360    │ ~396.13MiB 1224k m3u8_n │ avc1.42c01e 1224k video only
sec-540                         mp4 960x540    │ ~571.11MiB 1765k m3u8_n │ avc1.4d401f 1765k video only
hls-2016                        mp4 960x540    │ ~652.43MiB 2016k m3u8_n │ avc1.4d401f 2016k video only
sec-720-0                       mp4 1280x720   │ ~796.94MiB 2463k m3u8_n │ avc1.4d401f 2463k video only
hls-2730                        mp4 1280x720   │ ~883.22MiB 2730k m3u8_n │ avc1.4d401f 2730k video only
sec-720-1                       mp4 1280x720   │ ~  1.07GiB 3380k m3u8_n │ avc1.4d401f 3380k video only
hls-3667                        mp4 1280x720   │ ~  1.16GiB 3667k m3u8_n │ avc1.4d401f 3667k video only
sec-1080                        mp4 1920x1080  │ ~  1.88GiB 5939k m3u8_n │ avc1.4d401f 5939k video only
```

Feel free to make or suggest any changes - I just want to make my findings available to others

### CODE REVIEW

File: yt_dlp/extractor/cbc.py [hunk 1]
```diff
@@ -335,6 +371,7 @@ def _real_extract(self, url):
 
         formats = self._extract_m3u8_formats(m3u8_url, video_id, m3u8_id='hls')
         self._remove_duplicate_formats(formats)
+        self._find_secret_formats(formats, video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/cbc.py:374]:
```suggestion
        formats.extend(self._find_secret_formats(formats, video_id))
```
(see the other suggestions)

File: yt_dlp/extractor/cbc.py [hunk 2]
```diff
@@ -313,6 +314,41 @@ def _real_initialize(self):
...
+                    height = int(video_quality.attrib['MaxHeight'])
+                    bitrate = int(video_quality.attrib['Bitrate'])
+
+                    new_f = copy.deepcopy(base_f)
+                    new_f['format_id'] = f"sec-{height}"
+                    new_f['url'] = re.sub('(QualityLevels\\()\\d+(\\))', f'\\g<1>{bitrate}\\2', base_url)
+                    new_f['tbr'] = bitrate / 1000.0
+                    new_f['width'] = width
+                    new_f['height'] = height
+                    formats.append(new_f)
```
Comment (pukkandan) [inline yt_dlp/extractor/cbc.py:350]:
```suggestion
        """ Find a valid video url and convert it to the secret variant """
        base_format = next((f for f in formats if f.get('height')), None)
        if not base_format:
            return

        base_url = re.sub(r'(Manifest\(.*?),filter=[\w-]+(.*?\))', r'\1\2', base_format['url'])
        url = re.sub(r'(Manifest\(.*?),format=[\w-]+(.*?\))', r'\1\2', base_url)

        secret_xml = self._download_xml(url, video_id, note='Downloading secret XML', fatal=False)
        if not secret_xml:
            return

        for child in secret_xml:
            if child.attrib.get('Type') != 'video':
                continue
            for video_quality in child:
                bitrate = int_or_none(video_quality.attrib.get('Bitrate'))
                if not bitrate or 'Index' not in video_quality.attrib:
                    continue
                height = int_or_none(video_quality.attrib.get('MaxHeight'))

                yield {
                    **base_format,
                    'format_id': join_nonempty('sec', height),
                    'url': re.sub(r'(QualityLevels\()\d+(\))', fr'\1{bitrate}\2', base_url)
                    'width': int_or_none(video_quality.attrib.get('MaxWidth')),
                    'tbr': bitrate / 1000.0,
                    'height': height
                }
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
please add the necessary imports and test I didnt break anything

File: yt_dlp/extractor/cbc.py [hunk 3]
```diff
@@ -313,6 +314,37 @@ def _real_initialize(self):
...
+                    continue
+                height = int_or_none(video_quality.attrib.get('MaxHeight'))
+
+                yield {
+                    **base_format,
+                    'format_id': join_nonempty('sec', height),
+                    'url': re.sub(r'(QualityLevels\()\d+(\))', fr'\1{bitrate}\2', base_url)
+                    'width': int_or_none(video_quality.attrib.get('MaxWidth')),
+                    'tbr': bitrate / 1000.0,
+                    'height': height
```
Comment (pukkandan) [inline yt_dlp/extractor/cbc.py:345]:
```suggestion
                    **base_format,
                    'format_id': join_nonempty('sec', height),
                    'url': re.sub(r'(QualityLevels\()\d+(\))', fr'\1{bitrate}\2', base_url),
                    'width': int_or_none(video_quality.attrib.get('MaxWidth')),
                    'tbr': bitrate / 1000.0,
                    'height': height,
```

File: yt_dlp/extractor/cbc.py [hunk 4]
```diff
@@ -313,6 +314,37 @@ def _real_initialize(self):
             return
         self._claims_token = self._downloader.cache.load(self._NETRC_MACHINE, 'claims_token')
 
+    def _find_secret_formats(self, formats, video_id):
+        """ Find a valid video url and convert it to the secret variant """
+        base_format = next((f for f in formats if f.get('height')), None)
+        if not base_format:
+            return
```
Comment (pukkandan) [inline yt_dlp/extractor/cbc.py:321]: I don't understand why this is the selection criteria

Comment (DavidSkrundz) [reply yt_dlp/extractor/cbc.py:321]: It needs to select a video, and width/height was the first differentiator that I found. Looking at it again, `if f.get('vcodec') != 'none'` could also work

Comment (pukkandan) [reply yt_dlp/extractor/cbc.py:321]: please do that. `acodec/vcodec='none'` is the canonical way to identify video/audio-only formats

**Comment by mshiels** [comment]
Well the code may work, but for me nothing downloads, still working to grab log etc, but it's just ending in '400' errors.

**Comment by mshiels** [comment]
but maybe this will help - thinking regex problems!!

[debug] Invoking downloader on "https://cbcrcott-gem.akamaized.net/7faf85fd-b294-452a-9b76-a416a05b555b/CBC_CORONER_S04T01.ism/\<1>6041885)/Manifest(video,format=m3u8-aapl)"

vs a working one which still grabs 720p

[debug] Invoking downloader on "https://cbcrcott-gem.akamaized.net/7faf85fd-b294-452a-9b76-a416a05b555b/CBC_CORONER_S04T01.ism/QualityLevels(3411834)/Manifest(video,format=m3u8-aapl,filter=desktop)"

**Comment by nyuszika7h** [comment]
Yeah, found the mistake: https://github.com/yt-dlp/yt-dlp/commit/54c2521ca656f1a0897d21cbdf34cbad4e51735b#r61916098

**Comment by mshiels** [comment]
cool, so no need for detailed logs?

**Comment by DavidSkrundz** [comment]
Fixed in https://github.com/yt-dlp/yt-dlp/pull/2020, though for some reason I'm not experiencing issues with a missing `g`.  I'm curious as to what is the difference between our environments.  I'm running `python 3.10.0 (default, Oct 22 2021, 13:24:07) [Clang 13.0.0 (clang-1300.0.29.3)]` on macOS

**Comment by makeworld-the-better-one** [comment]
Nice work. Happy to see these secret formats don't even need an account (for videos that work without an account already).

**Pull request merged**